shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/datasource/state/DataSourceStateManager.java

```diff
@@ -0,0 +1,151 @@
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.infra.datasource.state;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.shardingsphere.infra.config.database.DatabaseConfiguration;
+import org.apache.shardingsphere.infra.datasource.state.exception.DataSourceStateException;
+
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Data source state manager.
+ */
+@Slf4j
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class DataSourceStateManager {
+    
+    private static final DataSourceStateManager INSTANCE = new DataSourceStateManager();
+    
+    private final Map<String, DataSourceState> dataSourceStates = new ConcurrentHashMap<>();
+    
+    private volatile boolean force;
+    
+    private volatile boolean initialized;
+    
+    /**
+     * Get data source state manager.
+     *
+     * @return data source state manager
+     */
+    public static DataSourceStateManager getInstance() {
+        return INSTANCE;
+    }
+    
+    /**
+     * Set data source states when bootstrap.
+     *
+     * @param databaseName database name
+     * @param dataSources data sources
+     * @param storageDataSourceStates storage node data source status
+     * @param force whether to force start
+     */
+    public void initStates(final String databaseName, final Map<String, DataSource> dataSources, final Map<String, DataSourceState> storageDataSourceStates, final boolean force) {
+        this.force = force;
+        dataSources.forEach((key, value) -> {
+            DataSourceState storageState = storageDataSourceStates.get(getCacheKey(databaseName, key));
+            if (DataSourceState.DISABLED == storageState) {
+                dataSourceStates.put(getCacheKey(databaseName, key), storageState);
+            } else {
+                try (Connection ignored = value.getConnection()) {
+                    dataSourceStates.put(getCacheKey(databaseName, key), DataSourceState.ENABLED);
+                } catch (final SQLException ex) {
+                    if (this.force) {
+                        log.error("Data source status unavailable, ignored with the -f parameter.", ex);
+                    } else {
+                        throw new DataSourceStateException("DataSourceState", 1, "Data source status unavailable.", ex);
+                    }
+                }
+            }
+        });
+        initialized = true;
+    }
+    
+    /**
+     * Get enabled data sources.
+     *
+     * @param databaseConfigs database config
+     * @return enabled data sources
+     */
+    public Collection<DataSource> getEnabledDataSources(final Map<String, ? extends DatabaseConfiguration> databaseConfigs) {
```

Comment:
databaseConfigs represents multiple logical libraries, is there a problem with the processing logic here?

Reply:
The unique data source state key is `databaseName + dataSourceName`.